### DISCUSSION TITLE
Implementation for Client Credentials Auth

### DISCUSSION THREAD

**Discussion by gthaka** [rhertogh/yii2-oauth2-server#4] [Q&A] [answered]
@rhertogh Do you have a sample implementation example of the Client Credentials grant type?

**Comment by rhertogh** [comment, marked as answer]
Hi @gthaka,
The Client Credentials grant type is meant for so called "two-legged OAuth", in other words direct client to server authentication without any end user being involved. Is that what you're trying to achieve? And does your application currently use RBAC?

The sample app also includes a client with the Client Credentials. You can start it in the following way:
1. Clone the repo: `git clone https://github.com/rhertogh/yii2-oauth2-server.git`
2. Change the working directory: `cd yii2-oauth2-server`
3. Copy the sample environment file `cp .env.example .env`
4. Build the docker containers: `docker compose build`
5. Start the docker containers: `docker compose up -d`
6. Change the file owner group for the sample keys: `docker compose exec Yii2Oauth2Server chown :www-data sample/config/keys/public.key sample/config/keys/private.key`
7. Change the permissions for the sample keys: `docker compose exec Yii2Oauth2Server chmod 660 sample/config/keys/public.key sample/config/keys/private.key`
8. Install the dependencies: `docker compose exec Yii2Oauth2Server composer install`
9. Make the 'yii' command executable: `docker compose exec Yii2Oauth2Server chmod +x ./yii`
10. Run the migrations: `docker compose exec Yii2Oauth2Server yii migrate --interactive=0`

You should then be able to configure your client, for example in Postman:

Create a new Collection with the following setting:
| Setting | Value |
|----------|------|
| Type | `Oauth 2.0` |
| Add auth data to | `Request Header` |
| Grant Type | `Client Credentials` |
| Access Token URL | `http://localhost/oauth2/access-token` |
| Client ID | `sample-client-credentials-client` |
| Client Secret | `secret` |
| Client Authentication | `Send as Basic Auth header` |

![image](https://user-images.githubusercontent.com/1292337/187570165-5ddd2063-7680-4660-b3c9-0f3ee67ea035.png)

Click "Get New Access Token", you should now receive an access token.

To test the token you can create a new GET request with the URL `http://localhost/api/user/me`.

**Reply by gthaka** [reply]
Thanks @rhertogh . 
I understand what Client Credentials grant type is for (M2M).

I was successful in testing it out by following your instructions. I will definitely use it to work on my implementation.

Thanks. 👌💯
